projects/com.oracle.truffle.llvm.parser/src/com/oracle/truffle/llvm/parser/LazyToTruffleConverterImpl.java

```diff
@@ -109,21 +117,83 @@ public RootCallTarget convert() {
         LLVMSourceLocation location = method.getLexicalScope();
 
         List<LLVMStatementNode> copyArgumentsToFrame = copyArgumentsToFrame(frame);
         LLVMStatementNode[] copyArgumentsToFrameArray = copyArgumentsToFrame.toArray(new LLVMStatementNode[copyArgumentsToFrame.size()]);
-        LLVMExpressionNode body = runtime.getNodeFactory().createFunctionBlockNode(frame.findFrameSlot(LLVMUserException.FRAME_SLOT_ID), visitor.getBlocks(), uniquesRegion.build(),
-                        nullableBeforeBlock, nullableAfterBlock, location, copyArgumentsToFrameArray);
 
+        // create function body node
+        LLVMExpressionNode body;
+        if (cfg.isReducible()) {
+            List<LLVMStatementNode> nodes = new ArrayList<>(visitor.getBlocks());
+            body = resolveLoops(nodes, cfg, frame, nullableBeforeBlock, nullableAfterBlock,
+                            location, copyArgumentsToFrameArray, uniquesRegion);
+        } else {
+            body = runtime.getNodeFactory().createFunctionBlockNode(frame.findFrameSlot(LLVMUserException.FRAME_SLOT_ID), visitor.getBlocks(), uniquesRegion.build(), nullableBeforeBlock,
+                            nullableAfterBlock, location, copyArgumentsToFrameArray, null);
+        }
         RootNode rootNode = runtime.getNodeFactory().createFunctionStartNode(runtime.getContext(), body, method.getSourceSection(), frame, method, source, location);
         method.onAfterParse();
 
         return Truffle.getRuntime().createCallTarget(rootNode);
     }
 
+    private LLVMExpressionNode resolveLoops(List<LLVMStatementNode> nodes, LLVMControlFlowGraph cfg, FrameDescriptor frame, FrameSlot[][] nullableBeforeBlock,
+                    FrameSlot[][] nullableAfterBlock, LLVMSourceLocation location, LLVMStatementNode[] copyArgumentsToFrameArray, UniquesRegion uniquesRegion) {
+
+        if (cfg.getCFGLoops().size() > 0) {
+            loopSuccessorSlot = frame.addFrameSlot("__LLVMLoopSuccessor", FrameSlotKind.Int);
+        }
+
+        LLVMStatementNode[] nodeArray = nodes.toArray(new LLVMStatementNode[0]);
+
+        for (CFGLoop l : cfg.getCFGLoops()) {
+
+            int headerId = l.getHeader().id;
+
+            Integer[] indexMapping = new Integer[nodes.size()];
+            Arrays.fill(indexMapping, -1);
+
+            List<LLVMStatementNode> bodyNodes = new ArrayList<>();
+
+            // add header to body nodes
+            bodyNodes.add(nodes.get(headerId));
+            indexMapping[headerId] = 0;
+
+            // add body nodes
+            int i = 1;
+            for (CFGBlock b : l.getBody()) {
+                bodyNodes.add(nodeArray[b.id]);
+                indexMapping[b.id] = i++;
+            }
+
+            LLVMExpressionNode loopBody = runtime.getNodeFactory().createLoopDispatchNode(frame.findFrameSlot(LLVMUserException.FRAME_SLOT_ID), Collections.unmodifiableList(bodyNodes),
```

Comment:
`getSuccessorIDs` allocates a new array with the same content on each call. Since, I assume, the contents of the returned arrays are expected to be identical, you should only call the method once and pass the same array to `createLoop` and `createLoopDispatchNode`.